src/compiler/scala/tools/nsc/transform/async/AsyncPhase.scala

```diff
@@ -0,0 +1,287 @@
+          }
+          override protected def description: String = ""
+        }
+        val AsyncModule = packSym.moduleClass.newModule("Async")
+        val moduleClass = AsyncModule.moduleClass
+        moduleClass.setInfo(ClassInfoType(Nil, newScope, moduleClass))
+        AsyncModule.setInfoAndEnter(TypeRef(packSym.thisType, moduleClass, Nil))
+
+        def futureOfT(tparam: Symbol) = appliedType(rootMirror.getClassIfDefined("scala.concurrent.Future").tpeHK, tparam.tpeHK)
+
+        val Async_async = AsyncModule.moduleClass.newMethodSymbol(nme.async, newFlags = Flags.MACRO)
+        val Async_async_T = Async_async.newSyntheticTypeParam("T", 0L).setInfo(TypeBounds.empty)
+        val Async_async_body = Async_async.newSyntheticValueParam(Async_async_T.tpeHK, "body")
+        val Async_async_execContext = Async_async.newSyntheticValueParam(rootMirror.getClassIfDefined("scala.concurrent.ExecutionContext").tpeHK, "executionContext").setFlag(Flags.IMPLICIT)
+        Async_async.setInfoAndEnter(PolyType(Async_async_T :: Nil, MethodType(Async_async_body :: Nil, MethodType(Async_async_execContext :: Nil, futureOfT(Async_async_T)))))
+
+        val Async_await = AsyncModule.moduleClass.newMethodSymbol(nme.await)
+        val Async_await_T = Async_await.newSyntheticTypeParam("T", 0L).setInfo(TypeBounds.empty)
+        val Async_await_future = Async_await.newSyntheticValueParam(futureOfT(Async_await_T), "awaitable")
+        Async_await.setInfoAndEnter(PolyType(Async_await_T :: Nil, MethodType(Async_await_future :: Nil, Async_await_T.tpeHK)))
+        new AsyncSymbols(Async_async, Async_await)
+      case _ =>
+        val AsyncModule = rootMirror.getModuleIfDefined("scala.async.Async")
+        val Async_async = AsyncModule.map(async => definitions.getDeclIfDefined(async, nme.async))
+        val Async_await = AsyncModule.map(async => definitions.getDeclIfDefined(async, nme.await))
+        new AsyncSymbols(Async_async, Async_await)
+    }
+  }
+
+  override def newPhase(prev: scala.reflect.internal.Phase): StdPhase = {
+    new Phase(prev) {
+      override def init(): Unit = {
+        asyncSymbols // force
+      }
+    }
+  }
+  import treeInfo.Applied
+  def fastTrackEntry: (Symbol, PartialFunction[Applied, scala.reflect.macros.contexts.Context { val universe: self.global.type } => Tree]) =
+    (asyncSymbols.Async_async, {
+      // def async[T](body: T)(implicit execContext: ExecutionContext): Future[T] = macro ???
+      case app@Applied(_, _, List(asyncBody :: Nil, execContext :: Nil)) =>
+        c => c.global.async.macroExpansion.apply(c.callsiteTyper, asyncBody, execContext, asyncBody.tpe)
+    })
+
+  def newTransformer(unit: CompilationUnit): Transformer = new AsyncTransformer(unit)
+
+  // TOOD: figure out how to make the root-level async built-in macro sufficiently configurable:
+  //       replace the ExecutionContext implicit arg with an AsyncContext implicit that also specifies the type of the Future/Awaitable/Node/...?
```

Comment:
is this done now with the latest rework?

Reply:
Not exactly. Our current means of customisation is that each different future type still needs to provide a separate async macro.

To address this comment we would want to build upon the last commits here with something like:

```scala
trait Awaitable[F[_]] {
  type R[A]
  protected def completeFailure(t: Throwable): Unit
  protected def completeSuccess(value: AnyRef): Unit
  protected def onComplete(f: F[AnyRef]): Unit
  protected def getCompleted(f: F[AnyRef]): R[AnyRef]
  protected def tryGet(tr: R[AnyRef]): AnyRef
}

implicit def ScalaConcurrentIsAwaitable(implicit ec: ExecutionContext): Awaitable[Future[_]] { type R[A] = Try[A]; ... }

def async[F[_]: Awaitable, T](f: F[T]): T = macro asyncImpl
```